kivy/clock.py

```diff
@@ -626,36 +618,71 @@ def unschedule(self, callback, all=True):
             callback.cancel()
         else:
             if all:
-                for ev in self._events[_hash(callback)][:]:
+                events = []
+                self._lock_acquire()
+
+                ev = self._root_event
+                while ev is not None:
                     if ev.get_callback() == callback:
-                        ev.cancel()
+                        events.append(ev)
+                    ev = ev.next
+
+                self._lock_release()
+                for ev in events:
+                    ev.cancel()
             else:
-                for ev in self._events[_hash(callback)][:]:
+                cancel = None
+                self._lock_acquire()
+
+                ev = self._root_event
+                while ev is not None:
                     if ev.get_callback() == callback:
-                        ev.cancel()
+                        cancel = ev
                         break
+                    ev = ev.next
+                self._lock_release()
+                if cancel is not None:
+                    cancel.cancel()
 
     def _release_references(self):
         # call that function to release all the direct reference to any
         # callback and replace it with a weakref
-        events = self._events
-        for events in self._events:
-            for event in events[:]:
-                if event.callback is not None:
-                    event.release()
+        events = []
+        self._lock_acquire()
+
+        ev = self._root_event
+        while ev is not None:
+            if ev.callback is not None:
+                events.append(ev)
+            ev = ev.next
+
+        self._lock_release()
+        for ev in events:
+            ev.release()
 
     def _process_events(self):
-        for events in self._events:
-            remove = events.remove
-            for event in events[:]:
-                # event may be already removed from original list
-                if event in events:
-                    event.tick(self._last_tick, remove)
+        acquire, release = self._lock_acquire, self._lock_release
+
+        acquire()
+        event = self._root_event
+        while event is not None:
+            self._next_event = event.next
+            release()
+
+            try:
+                event.tick(self._last_tick)
+            except:
+                raise
+            else:
+                acquire()
+            event = self._next_event
+        release()
 
     def _process_events_before_frame(self):
-        found = True
+        acquire, release = self._lock_acquire, self._lock_release
```

Comment:
just a "perf" notation, it's less bytecode to write 2 lines, instead of generating a tuple and unpack it here ^^